src/Compilers/CSharp/Portable/Syntax/NonRecursiveSyntaxRewriter.cs

```diff
@@ -0,0 +1,286 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph and
+    /// may replace or remove visited SyntaxNodes in depth-first order.
+    /// </summary>
+    public partial class NonRecursiveSyntaxRewriter : CSharpSyntaxVisitor<NonRecursiveSyntaxRewriter.Chunk>
+    {
+        public class Chunk
```

Comment:
Actually the public is inherited from CSharpSyntaxVisitor<T>. 
Would you prefer that I add a new Accept methods on each node?
BTW if you don't pass the function, does it mean that you would prefer I also add a new AcceptTransform method?

For using a struct (vs a class) for Chunk, I'm agree.